Python/_warnings.c

```diff
@@ -926,22 +929,20 @@ setup_context(Py_ssize_t stack_level,
          if (rc < 0)
             goto handle_error;
     }
-    else
-        Py_INCREF(*registry);
 
     /* Setup module. */
-    *module = _PyDict_GetItemWithError(globals, &_Py_ID(__name__));
-    if (*module == Py_None || (*module != NULL && PyUnicode_Check(*module))) {
-        Py_INCREF(*module);
-    }
-    else if (_PyErr_Occurred(tstate)) {
+    rc = PyDict_GetItemRef(globals, &_Py_ID(__name__), module);
+    if (rc < 0) {
         goto handle_error;
     }
-    else {
+    if (rc == 0) {
         *module = PyUnicode_FromString("<string>");
         if (*module == NULL)
             goto handle_error;
     }
+    else {
+        assert(Py_IsNone(*module) || PyUnicode_Check(*module));
```

Comment:
I think we need to handle the case where `*module` (i.e., `__name__`) is not `None` and not a unicode object because Python code can set `__name__` arbitrarily.

Reply:
4f3cd028b6bd1617e48ed49b500d7e3bca2aac27 should fix that case:

If None or a unicode object, return successfully; else decref and fall through to the no-key case.

Reply:
Note, there is no test case for this in the test suite.